src/core/search/base.h

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <ostream>
+#include <regex>
+#include <variant>
+#include <vector>
+
+#include "core/core_types.h"
+#include "core/string_map.h"
+
+namespace dfly::search {
+
+// Represents input to a search tree.
+struct SearchInput {
+  // Callback that's supplied with field values.
+  using FieldConsumer = std::function<bool(std::string_view)>;
+
+  virtual ~SearchInput() = default;
+
+  // Check if the given callback returns true on any of  the active fields.
+  virtual bool Check(FieldConsumer) = 0;
+
+  // Sets a single active field.
+  virtual void SelectField(std::string_view field) = 0;
+
```

Comment:
Not sure about the interface yet.

Virtual so its fully mockable. Otherwise search needs to include core structures like StringMap and redis skiplists (or should it?)

I also don't like the FieldConsumer function glue here, maybe have a separate virtual base for nodes that work directly with strings

Reply:
I'm not a fan of the fact that `input` is now a mutable argument to all expression checkers, but I don't have a better idea (nor do I fully understand the challenges here, tbh).
Since this is supposed to be a quick-N-dirty test ground, it's great to let it evolve as you go. If you do want to brainstorm about this - lmk, I'd be happy to!

Reply:
Agree. Yes, this part won't outlive the MVP, but lets still brainstorm quickly 🙂 

1. Mutable state

The only state we have is the field (and I'm almost sure there won't be more). So if your query has `@field:(subtree)`, then the whole subtree needs to be applied only to a specific field of the document, whereas by default they apply to all fields. I do this field shift with a single node that sets the active field.

The better idea for now would be most likely smth like this:

```c++

// Interface for accessing hset values with different data structures underneath.
struct HSetAccessor {
  using FieldConsumer = std::function<bool(std::string_view)>;
  virtual bool Check(FieldConsumer f, std::string_view active_field = {}) const;
};

struct SearchInput {
  const HSetAccessor* hset;
  std::string_view active_field;
};
```

So now its all const and state is decoupled from the interface. The downside I see in the future is that with explicit SelectField/ClearField the real implementation can store an iterator/pointer and look up this value again much faster, as it could be used in a large subtree. But it can also do it as an internal optimization, comparing last accessed values.

3. FieldConsumer

Some context about the future: Why not fill an container like vector<string_view>*

- The data structure can't always provide string_views to all its entries. With listpack some might be stored as real integers and need to be decoded to a temporary buffer (so we also cant return a single string_view, only if a buffer is passed)
- We probably don't want to iterate over all fields if we match on the first

What I don't like with the FieldConsumer is that:

- The iteration logic is handled by the data structure accessor
- The node has to wrap its logic into one more layer

So I'd make the HSetAccessor return just some kind of virtual iterator. C++ ones are a pain really. Go style isn't bad.

```c++
struct HSetIterator {
  virtual string_view Next();
  virtual bool HasNext():
}
```

So iteration logic can be performed by the nodes themselves... But it'll be the same for all of them. Keeping it in a middle-part like SearchInput as a helper requires one more virtual base.